airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.kt

```diff
@@ -459,6 +459,42 @@ open class S3StorageOperations(
         }
     }
 
+    override fun listExistingObjects(
+        namespace: String?,
+        streamName: String,
+        objectPath: String,
+        pathFormat: String
+    ): List<String> {
+        val regexFormat: Pattern =
+            Pattern.compile(getRegexFormat(namespace, streamName, pathFormat))
+        var keys = listOf<String>()
+        var objects: ObjectListing = listObjects(objectPath)
+        while (objects.objectSummaries.size > 0) {
```

Comment:
I feel like we're copying this loop in a few places (here, getStageGeneration in https://github.com/airbytehq/airbyte/pull/42543). Can we make a method like this?
```kotlin
fun forEachObject(namespace, streamName, objectPath, pathFormat, consume: S3ObjectSummary -> Unit) {
  while (objects.objectSumaries.size > 0) {
    objects.objectSummaries.filter {regex...}.forEach { consume(it) }
    if (objects.isTruncated) // ...
  }
}
```

and then we could do approximately these things:
```kotlin
fun listExistingObjects(...): List<String> {
  var keys = mutableListOf<String>()
  // technically kind of inefficient, but I assume we're handling at most a few hundred thousand objects?
  forEachObject(...) { keys.add(it.key) }
  return keys
}

fun getStageGeneration(...) {
  var lastModifiedObject: S3ObjectSummary? = null
  forEachObject(...) {
    if (
        lastModifiedObject == null ||
            descendingComparator.compare(lastModifiedObject, it) > 0
    ) {
        lastModifiedObject = it
    }
  }
  return lastModifiedObject.getMetadata(...)
}
```

Reply:
I was hoping you wouldn't notice 😄

Reply:
consolidated 3 usages. left 1 occurrence which I'm not sure is even working (unused in current code)